storage/wal/storage.go

```diff
@@ -0,0 +1,754 @@
+					Segment: r.Segment(),
+					Offset:  r.Offset(),
+				}
+			}
+		}
+	}()
+
+	for d := range decoded {
+		switch v := d.(type) {
+		case []record.RefSeries:
+			for _, entry := range v {
+				// If this is a new series, create it in memory. If we never read in a
+				// sample for this series, its timestamp will remain at 0 and it will
+				// be deleted at the next GC.
+				if s.series.GetByID(entry.Ref) == nil {
+					series := &memSeries{ref: entry.Ref, lset: entry.Labels, lastTs: 0}
+					s.series.Set(entry.Labels.Hash(), series)
+
+					s.metrics.numActiveSeries.Inc()
+
+					if entry.Ref > lastRef {
+						lastRef = entry.Ref
+					}
+				}
+			}
+
+			//nolint:staticcheck
+			seriesPool.Put(v)
+		case []record.RefSample:
+			for _, entry := range v {
+				// Update the lastTs for the series based
+				series := s.series.GetByID(entry.Ref)
+				if series == nil {
+					level.Warn(s.logger).Log("msg", "found sample referencing non-existing series, skipping")
+					continue
+				}
+
+				series.Lock()
+				if entry.T > series.lastTs {
+					series.lastTs = entry.T
+				}
+				series.Unlock()
+			}
+
+			//nolint:staticcheck
+			samplesPool.Put(v)
+		default:
+			panic(fmt.Errorf("unexpected decoded type: %T", d))
+		}
+	}
+
+	s.nextRef.Store(lastRef)
+
+	select {
+	case err := <-errCh:
+		return err
+	default:
+		if r.Err() != nil {
+			return errors.Wrap(r.Err(), "read records")
+		}
+		return nil
+	}
+}
+
+func (s *Storage) run() {
+	defer close(s.donec)
+
+Loop:
+	for {
+		select {
+		case <-s.stopc:
+			break Loop
+		case <-time.After(s.opts.TruncateFrequency):
+			ts := s.rs.LowestSentTimestamp() - s.opts.MinWALTime.Milliseconds()
+			if ts < 0 {
+				ts = 0
+			}
+
+			// Clamp timestamp to be the maximum lifetime of WAL data.
+			if maxTS := timestamp.FromTime(time.Now().Add(-s.opts.MaxWALTime)); ts < maxTS {
+				ts = maxTS
+			}
+
+			level.Debug(s.logger).Log("msg", "truncating the WAL", "ts", ts)
+			if err := s.truncate(ts); err != nil {
+				level.Warn(s.logger).Log("msg", "failed to truncate WAL", "err", err)
+			}
+		}
+	}
+
+	if s.opts.WriteStaleOnShutdown {
+		if err := s.writeStalenessMarkers(); err != nil {
+			level.Warn(s.logger).Log("msg", "failed to write staleness markers", "err", err)
+		}
+	}
+}
+
+func (s *Storage) writeStalenessMarkers() error {
+	var (
+		lastTs int64
+		app    = s.Appender(context.Background())
+	)
+
+	for refLock := 0; refLock < s.series.size; refLock++ {
+		s.series.locks[refLock].RLock()
+
+		for _, series := range s.series.series[refLock] {
+			series.Lock()
+
+			hashLock := int(series.lset.Hash()) & (s.series.size - 1)
+			if refLock != hashLock {
+				s.series.locks[hashLock].RLock()
+			}
+
+			ts := timestamp.FromTime(time.Now())
+			_, err := app.Append(series.ref, series.lset, ts, math.Float64frombits(value.StaleNaN))
+			if err != nil {
+				return errors.Wrap(err, "append staleness marker")
+			}
+
+			// Remove millisecond precision; the remote write timestamp we
```

Comment:
what timestamp is second precision??

Reply:
https://github.com/prometheus/prometheus/blob/b50f9c1c8486b403c0299e84498eeec73f475e1a/storage/remote/queue_manager.go#L1040 IIUC, `queue_highest_sent_timestamp_seconds` is stored in seconds and not ms

Reply:
This code should go away anyway since we should not "invent" staleness markers.